### PR TITLE
Fix unpacking of artifact config

### PR DESCRIPTION
**Pull request by michalpristas**
Based on #686 added workaround for the issue when inline transport settings prevents other options in a artifact config to be unpacked properly.

This divide and conquer approach divides config into 2 struct one for transport one for artifact specific configs and each of them is reloaded separately.

Keeps source uri in a separate func as we need to handle case when fleet uses different kind of keywords.

Fixes: #408

Playground with repro: https://go.dev/play/p/V3BGRFruEm4

### CODE REVIEW

**Review by ph — APPROVED** [review]
@michalpristas Concerning the test are all the path tested?

**Comment by michalpristas** [comment]
@ph actually no, i was thinking about some more usecases while watching tv before sleep :D

**Comment by michalpristas** [comment]
do not merge i know what's wrong with original unpacking issue

**Comment by ph** [comment]
@michalpristas adding the reproducible case in this PR would be useful too for later reference.

File: internal/pkg/artifact/config.go [hunk 1]
```diff
@@ -74,17 +111,19 @@ func (r *Reloader) Reload(rawConfig *config.Config) error {
...
 		// fleet configuration takes precedence
-		newSourceURI = cfg.FleetSourceURI
-	} else if cfg.SourceURI != "" {
-		newSourceURI = cfg.SourceURI
+		newSourceURI = fleetURI
+	} else if sourceURI := strings.TrimSpace(cfg.SourceURI); sourceURI != "" {
+		newSourceURI = sourceURI
 	}
 
+	newSourceURI = strings.TrimSpace(newSourceURI)
```
Comment (ph) [inline internal/pkg/artifact/config.go:125]: Question?

Are we validating anywhere that the source URI is valid? By valid I mean the format is valid and not reaching the host?

**Review by ph — APPROVED** [review]
Changes looks fine to me

File: internal/pkg/artifact/config.go (hunk 1 above)
Comment (michalpristas) [reply internal/pkg/artifact/config.go:125]: no we dont do that download will fail when trying to use something not usable

File: internal/pkg/artifact/config.go [hunk 2]
```diff
@@ -64,6 +65,42 @@ func NewReloader(cfg *Config, log *logger.Logger) *Reloader {
 }
 
 func (r *Reloader) Reload(rawConfig *config.Config) error {
+	if err := r.reloadConfig(rawConfig); err != nil {
+		return errors.New(err, "failed to reload source URI")
```
Comment (aleksmaus) [inline internal/pkg/artifact/config.go:69]: "failed to reload source URI" ?
did you mean something like "failed to reload config"? cause the next error bellow has identical message, would not be able to tell one from another

File: internal/pkg/artifact/config.go [hunk 3]
```diff
@@ -64,6 +65,42 @@ func NewReloader(cfg *Config, log *logger.Logger) *Reloader {
...
+}
+
+func (r *Reloader) reloadConfig(rawConfig *config.Config) error {
+	type reloadConfig struct {
+		C *Config `json:"agent.download" config:"agent.download"`
+	}
+	tmp := &reloadConfig{
+		C: DefaultConfig(),
+	}
+	if err := rawConfig.Unpack(&tmp); err != nil {
```
Comment (aleksmaus) [inline internal/pkg/artifact/config.go:86]: is it possible tmp.C is nil after this? do we need the nil check for that?

File: internal/pkg/artifact/config.go [hunk 4]
```diff
@@ -74,17 +111,19 @@ func (r *Reloader) Reload(rawConfig *config.Config) error {
 	}
 	cfg := &reloadConfig{}
 	if err := rawConfig.Unpack(&cfg); err != nil {
-		return errors.New(err, "failed to unpack config during reload")
+		return err
```
Comment (aleksmaus) [inline internal/pkg/artifact/config.go:114]: what is the benefit of removing the additional context from the the error here? it was wrapping the error before the change

File: internal/pkg/artifact/config.go (hunk 1 above)
Comment (aleksmaus) [inline internal/pkg/artifact/config.go:125]: double trimming? newSourceURI is already TrimSpace-d

**Review by aleksmaus — COMMENTED** [review]
maybe worth addressing a couple of places, left comments.

File: internal/pkg/artifact/config.go (hunk 3 above)
Comment (michalpristas) [reply internal/pkg/artifact/config.go:86]: no we init with default and upack replaces values inside. nil C should not happen

**Pull request merged**